File: 1/2020.10.03/367.Valid Perfect Square/python/solution.py
class Solution:
    def isPerfectSquare(self, num: int) -> bool:
        # 牛顿迭代法
        if num < 2:
            return True
        x = num // 2
        # x * x 容易超出范围
        # while x * x > num:
        while x > num // x:
            x = (x + num // x) // 2
        return x * x == num

File: 1/2020.10.03/367.Valid Perfect Square/java/solution.java
class Solution {
    public boolean isPerfectSquare(int num) {
        // binary search
        if (num < 2) {
            return true;
        }
        long left = 2, right = num / 2;
        while (left <= right) {
            long x = left + (right - left) / 2;
            long multy = x * x;
            if (multy == num) {
                return true;
            } else if (multy < num) {
                left = x + 1;
            } else {
                right = x - 1;
            }
        }
        return false;
    }
}
